### PR TITLE
/ui: add `FontsLoader` component

### LINKED ISSUE #387: /ui: add `FontsLoader`
A component that lets you load fonts via Javascript.

As usual we should also use it in the doc site to check that it works correctly.

### PR DESCRIPTION
**Pull request by saabi**
Closes #387

Also:
- fix `StorageIO` bug occurring when component is unmounted

### CODE REVIEW

File: packages/docs/site/src/routes/_layout.svelte [hunk 1]
```diff
@@ -3,33 +3,45 @@
...
 	import Nav from 'app/components/Nav.svelte';
+	import {fontFamilies} from 'app/config';
 
 	export let segment;
 
+	let fontLoadStatus;
 	let showA11yMenu;
+
+	// $: console.log('typeface', $_a11ySettings.typeface.value);
+	// $: console.log('font families', $_fontFamilyNames);
```
Comment (mindrones) [inline packages/docs/site/src/routes/_layout.svelte]: leftovers

File: packages/components/ui/src/drivers/fonts/README.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+{#if fontLoadStatus?.isFirstLoaded}
+	<div>Content that depend on the first family being loaded.</div>
+{/if}
+```
+
+For a full implementation example please refer to
+`packages/docs/site/src/routes/_layout.svelte`.
+
+Props:
+- `FontsLoader`
```
Comment (mindrones) [inline packages/components/ui/src/drivers/fonts/README.md]: I guess this line is unnecessary

File: packages/components/ui/src/drivers/fonts/README.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	- `families`: An array of objects describing the families and typefaces to load.
```
Comment (mindrones) [inline packages/components/ui/src/drivers/fonts/README.md]: I think we need to explain:
- the format of the objects `{family, faces}`
- that not providing faces means assuming a system font (e.g. 'Courier New')

File: packages/components/ui/src/drivers/fonts/FontsLoader.svelte [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+	export let families = {};
+	export let firstFamilyToLoad = null;
+	export let status = $_status;
+
+	$: if (isClientSide) {
+		$_fontFamilies = families;
+		$_status.isDone  = false;
+	};
+	$: isClientSide && ($_firstFamilyToLoad = firstFamilyToLoad);
+	$: isClientSide && loadFamilies($_runtimeFamilies);
```
Comment (mindrones) [inline packages/components/ui/src/drivers/fonts/FontsLoader.svelte]: Can we move these two within the `if` above?

File: packages/components/ui/src/drivers/fonts/FontsLoader.svelte [hunk 5]
```diff
@@ -0,0 +1,24 @@
+<script>
+	import * as _ from 'lamb';
```
Comment (mindrones) [inline packages/components/ui/src/drivers/fonts/FontsLoader.svelte]: unused import

File: packages/components/ui/src/drivers/fonts/FontsLoader.svelte [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+	export let families = {};
+	export let firstFamilyToLoad = null;
+	export let status = $_status;
+
+	$: if (isClientSide) {
+		$_fontFamilies = families;
+		$_status.isDone  = false;
+	};
+	$: isClientSide && ($_firstFamilyToLoad = firstFamilyToLoad);
```
Comment (mindrones) [inline packages/components/ui/src/drivers/fonts/FontsLoader.svelte]: For `$_fontFamilies = families;` and `$_firstFamilyToLoad = firstFamilyToLoad` I'm getting `no-unused-vars` so I'd prefer using `store.set(value)` instead

File: packages/components/ui/src/drivers/fonts/stores.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
+import * as _ from 'lamb';
+import {makeMergeAppliedFnMap} from '@svizzle/utils';
```
Comment (mindrones) [inline packages/components/ui/src/drivers/fonts/stores.js]: unused import

File: packages/components/ui/src/io/storage/StorageIO.svelte [hunk 8]
```diff
@@ -40,7 +40,7 @@
 		_store.subscribe(updateDb);
 
 		return () => {
-			database.removeListener?.(key, setValue);
+			database.removeListener?.(key, updateStore);
```
Comment (mindrones) [inline packages/components/ui/src/io/storage/StorageIO.svelte]: [can't comment on line 24 as it's not part of the PR]

Re:

```
const updateStore = newValue => {
	// Current value in storage may be newer than value from event.
	newValue = database.getValue(key) || defaultValue;
	_store.set(newValue);
};
```

it's giving me a `no-param-reassign` eslint error, and I agree that reassigning an input var would be wrong, so I'd do something like

```
const updateStore = () => {
	// Current value in storage may be newer than value from event.
	const newValue = database.getValue(key) || defaultValue;
	_store.set(newValue);
};
```

as we don't seem to use `newValue` which seems to me an logical error?

File: packages/components/ui/src/drivers/fonts/stores.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+);
+export const _runtimeFamilies = derived(
+	[_fontFamilies, _firstFamilyToLoad],
+	([fontFamilies, firstFamilyToLoad]) => _.pipe([
+		_.collect([
+			_.filterWith(isFirstToLoad(firstFamilyToLoad)),
+			_.filterWith(_.not(isFirstToLoad(firstFamilyToLoad)))
+		]),
+	_.flatten,
+	getRuntimeFamilies
```
Comment (mindrones) [inline packages/components/ui/src/drivers/fonts/stores.js]: bad indentation

File: packages/docs/site/src/routes/_layout.svelte [hunk 10]
```diff
@@ -3,33 +3,41 @@
...
-				'Open Dyslexia'
-			],
-		}
-	}}
+defaults={{
+	typeface: {
+		defaultValue: fontFamilyNames,
+		values: fontFamilyNames,
+	}
+}}
```
Comment (mindrones) [inline packages/docs/site/src/routes/_layout.svelte]: bad indentation

File: packages/components/ui/src/drivers/fonts/stores.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+));
+const isFirstToLoad = firstFamilyToLoad => _.pipe([
+	_.getKey('family'),
+	_.is(firstFamilyToLoad)
+]);
+
+const getRuntimeFamily = _.adapter([
+	_.casus(_.hasKey('faces'), _.pipe([
+		_.collect([
+			_.getKey('family'),
```
Comment (mindrones) [inline packages/components/ui/src/drivers/fonts/stores.js]: `_.getKey('family')` is defined multiple times: we need to define it once and reuse it

File: packages/components/ui/src/drivers/fonts/stores.js [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+export const _fontFamilies = writable({});
+export const _status = writable({
+	isFirstLoaded: false,
+	isDone: false
+})
+
+// deriveds
+export const _fontFamilyNames = derived(
+	_fontFamilies,
+	_.mapWith(_.getKey('family'))
```
Comment (mindrones) [inline packages/components/ui/src/drivers/fonts/stores.js]: (1) 

I would have a new file `packages/components/ui/src/drivers/fonts/utils.js` with:

```
export const getFamily = _.getKey('family');
export const getFamilyNames = _.mapWith(getFamily);
```

and any utility we use here, so that we would import `getFamilyNames` from `@svizzle/ui/src/drivers/fonts/utils` in the app.

File: packages/docs/site/src/node_modules/app/config.js [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+				descriptors: {
+					weight: 700,
+					style: 'italic'
+				}
+			}
+		]
+	}
+];
+
+export const fontFamilyNames = fontFamilies.map(fontFamily => fontFamily.family);
```
Comment (mindrones) [inline packages/docs/site/src/node_modules/app/config.js]: I'd reuse `getFamilyNames` from `@svizzle/ui/src/drivers/fonts/utils`, see (1)

**Review by mindrones — CHANGES_REQUESTED** [review]
First pass

File: packages/components/ui/src/drivers/fonts/FontsLoader.svelte (hunk 4 above)
Comment (saabi) [reply packages/components/ui/src/drivers/fonts/FontsLoader.svelte]: Unfortunately, moving them within the `if` clause will cause them to be executed more times than needed and load fonts from the network every time a new font is selected.

Comment (saabi) [reply packages/components/ui/src/drivers/fonts/FontsLoader.svelte]: Maybe we should add a comment explaining this behaviour.

File: packages/components/ui/src/drivers/fonts/README.md [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+{#if fontLoadStatus?.isFirstLoaded}
+	<div>Content that depend on the first family being loaded.</div>
+{/if}
+```
+
+For a full implementation example please refer to
+`packages/docs/site/src/routes/_layout.svelte`.
+
+Props:
+- `families`: An array of objects of shape `{family: string, faces: {src: string, descriptors: []}}` describing the families and typefaces to load. For information on the `descriptors` object please refer to https://developer.mozilla.org/en-US/docs/Web/API/FontFace/FontFace.
```
Comment (mindrones) [inline packages/components/ui/src/drivers/fonts/README.md]: I think we still need to explain that not providing `faces` means assuming a system font (e.g. 'Courier New')

File: packages/docs/site/src/node_modules/app/config.js [hunk 15]
```diff
@@ -0,0 +1,78 @@
+import {getFamilyNames} from "@svizzle/ui/src/drivers/fonts/utils";
```
Comment (mindrones) [inline packages/docs/site/src/node_modules/app/config.js]: We need to use single apices here

File: packages/components/ui/src/drivers/fonts/utils.js [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+
+const ungroupFamily = (family, faces) => faces.map(face => new FontFace(
+	family,
+	face.src,
+	face.descriptors
+));
+export const isFirstToLoad = firstFamilyToLoad => _.pipe([
+	getFamily,
+	_.is(firstFamilyToLoad)
+]);
```
Comment (mindrones) [inline packages/components/ui/src/drivers/fonts/utils.js]: I'd use: 

```js
export const isFamilyEqualTo = family => _.pipe([getFamily, _.is(family)]);
```

or

```js
export const isFamilyEqualTo = isKeyValue(['family', family]);
```

File: packages/components/ui/src/drivers/fonts/README.md [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+- `firstFamilyToLoad`: If set, this property informs the component what family of typefaces to load first
```
Comment (mindrones) [inline packages/components/ui/src/drivers/fonts/README.md]: We need to specify that it's a font family name, a string

File: packages/components/ui/src/drivers/fonts/README.md [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+- `status`: This property can be bound to for receiving the status of the font loading process.
```
Comment (mindrones) [inline packages/components/ui/src/drivers/fonts/README.md]: We need to specify the properties of the object and their meaning

File: packages/components/ui/src/drivers/fonts/stores.js [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+
+import {
+	getFamilyNames,
+	isFirstToLoad,
+	getRuntimeFamilies
+} from './utils';
+
+// main stores
+export const _firstFamilyToLoad = writable('');
+export const _fontFamilies = writable({});
```
Comment (mindrones) [inline packages/components/ui/src/drivers/fonts/stores.js]: this is an array

```suggestion
export const _fontFamilies = writable([]);
```

File: packages/components/ui/src/drivers/fonts/FontsLoader.svelte [hunk 20]
```diff
@@ -0,0 +1,22 @@
...
+	import {isClientSide} from '../../utils/env';
+	import {
+		_firstFamilyToLoad,
+		_fontFamilies,
+		_runtimeFamilies,
+		_status,
+		loadFamilies
+	} from './stores';
+
+	export let families = {};
```
Comment (mindrones) [inline packages/components/ui/src/drivers/fonts/FontsLoader.svelte]: this is an array

```suggestion
	export let families = [];
```

File: packages/components/ui/src/drivers/fonts/stores.js [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+export const _runtimeFamilies = derived(
+	[_fontFamilies, _firstFamilyToLoad],
+	([fontFamilies, firstFamilyToLoad]) => _.pipe([
+		_.collect([
+			_.filterWith(isFirstToLoad(firstFamilyToLoad)),
+			_.filterWith(_.not(isFirstToLoad(firstFamilyToLoad)))
+		]),
+		_.flatten,
+		getRuntimeFamilies
+	])(fontFamilies));
```
Comment (mindrones) [inline packages/components/ui/src/drivers/fonts/stores.js]: I think this:

```js
_.collect([
	_.filterWith(isFirstToLoad(firstFamilyToLoad)),
	_.filterWith(_.not(isFirstToLoad(firstFamilyToLoad)))
]),
_.flatten,
```

could be reduced to 

```js
_.sortWith([_.sorterDesc(isFirstToLoad(firstFamilyToLoad))])
```

But I'd prefer something like (2):
- note that arguments are sorted alphabetically
- I'm using `isFamilyEqualTo` as suggested elsewhere in the PR
- the utils that weer used would be unused with the code below so they'd need to be deleted from `utils.js`

(2)

```suggestion
export const _runtimeFamilies = derived(
	[_firstFamilyToLoad, _fontFamilies],
	([firstFamilyToLoad, fontFamilies]) => {
		const isFirstFamily = isFamilyEqualTo(firstFamilyToLoad);

		const runtimeFamilies = _.reduce(
			fontFamilies,
			(acc, {family, faces}) => {
				const fontFaces = faces.map(
					({src, descriptors}) => new FontFace(family, src, descriptors)
				);
				const isFirstFontFace = isFirstFamily(family);

				if (isFirstFontFace) {
					acc.unshift(fontFaces)
				} else {
					acc.push(fontFaces)
				}

				return acc;
			},
			[]
		);

		return runtimeFamilies;
	}
);
```

**Review by mindrones — CHANGES_REQUESTED** [review]
pass two

File: packages/components/ui/src/drivers/fonts/FontsLoader.svelte (hunk 4 above)
Comment (saabi) [reply packages/components/ui/src/drivers/fonts/FontsLoader.svelte]: Since we're not using caching, these should remain as separate reactive statements for now.

**Review by mindrones — APPROVED** [review]
Great job! 👌
